decryptor/postgresql/packet_handler.go

```diff
@@ -427,100 +444,159 @@ const WithoutMessageType = 0
+// | StartupMessage      | F    | Int32(len) || Int32(196608)   |
+// | Sync                | F    | Byte1('S') || Int32(len)      |
+// | Terminate           | F    | Byte1('X') || Int32(len)      |
+// ```
+//
+// Startup message can only be received as first message after connection is
+// established. This fact allows to distinguish which format to parse.
+// If connection is established, and first message is not startup one, this
+// function returns ErrUnsupportedPacketType.
+// If startup message is already received, but then comes a message with unknown
+// format, it would be parsed as a general message with unknown tag.
+//
+// https://www.postgresql.org/docs/current/static/protocol-message-formats.html
 func (packet *PacketHandler) ReadClientPacket() error {
+	if packet.started {
+		return packet.readGeneralPacket()
+	}
+	// https://www.postgresql.org/docs/current/protocol-flow.html
+	// First packet should always be StartupMessage/SSLRequest/GSSENCRequest
+	return packet.readStartupPacket()
+}
+
+// Tries to read one of the startup packets:
+// - StartupMessage
+// - SSLRequest
+// - CancelRequest
+// - GSSENCRequest
+//
+// Due to historical reasons, all startup messages have the following format:
+//
+//     [4-byte length] [4-byte tag] [payload...]
+//
+// If the packet cannot be parsed as a startup packets, the ErrUnsupportedPacketType
+// is returned.
+//
+// Source: https://www.postgresql.org/docs/current/protocol-flow.html
+func (packet *PacketHandler) readStartupPacket() error {
 	packet.Reset()
-	// 8 bytes because startup/ssl/cancel messages has at least 8 bytes
+	// 8 bytes because all startup messages has at least 8 bytes
 	packetBuf := make([]byte, 8)
 	packet.messageType[0] = WithoutMessageType
-	// any message has at least 5 bytes: TypeOfMessage(1) + Length(4) or 8 bytes of special messages
+
+	n, err := io.ReadFull(packet.reader, packetBuf)
+	if err := base.CheckReadWrite(n, 8, err); err != nil {
+		packet.logger.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorDecryptorReadPacket).WithError(err).Debugln("Can't read 8 bytes of Startup/SSLRequest packet")
+		return err
+	}
+
+	switch {
+	case bytes.Equal(StartupRequest, packetBuf[4:8]):
+	case bytes.Equal(SSLRequestHeader, packetBuf[:8]):
+	case bytes.Equal(CancelRequestHeader, packetBuf[:8]):
+	case bytes.Equal(GSSENCRequestHeader, packetBuf[:8]):
+	default:
+		return ErrUnsupportedPacketType
+	}
+
+	packet.setDataLengthBuffer(packetBuf[:4])
+
+	// We read 8 bytes: 4-byte length and 4-byte tag.
+	// Return tag to the descriptionBuf
+	n, err = packet.descriptionBuf.Write(packetBuf[4:])
+	if err := base.CheckReadWrite(n, 4, err); err != nil {
+		return err
+	}
+
+	// we read 4 bytes before. decrease before call readData because it read exactly as dataLength
+	packet.dataLength -= 4
+
+	if err := packet.readData(false); err != nil {
+		return err
+	}
+	// restore correct value
+	packet.dataLength += 4
+
+	packet.started = true
```

Comment:
https://github.com/cossacklabs/acra/blob/0.92.0/decryptor/postgresql/pg_decryptor.go#L560

we should change this flag here too. When app connects without TLS, asks TLS and postgresql accepts it, then they establish TLS connection and sends startup message again. 
hm, maybe we should add integration test that will connect like that. When tries without TLS but works with TLS too. 
to do it app should use `sslmode=allow`

Reply:
also you can check manually that it fails. try to connect with psql and such sslmode through acra to postgresql that configured to use ssl. you can use our image used in postgresql-job in `.circleci/config.yml` and certificates in `tests/ssl/acra-client/*` and `tests/ssl/ca/ca.crt`

Reply:
yeah, you are right. missed that packetHandler responsible for parsing and processing first packet now. Last note, can you remove redundant `firstByte` variable? and it will be enough)